Automattic/woocommerce-payments
# Improve disputed order notice load time by directly fetching dispute rather than charge

Fixes #7163
Fixes #7076

### Changes proposed in this Pull Request

This PR is a proposal to improve the performance of the disputed order notice by doing the following:
1. When a dispute created or closed webhook is received
	1. Store the dispute ID as order meta
1. On order screen load
	1. Get the dispute ID from the order meta
		2. If no dispute ID order meta found, request the disputes API filtered by order_number to check for disputes created/received before this PR's release
	4. If the order's dispute ID exists, fetch the single dispute's data during page load (PHP)
	5. Dispute data is then provided to JS immediately on page load and the notice is rendered

* This results in an average **13.5% improvement in rendering time of this notice (refresh → notice visible)** on a WooExpress environment: _see metrics below_.
* Additionally, **testing WCCOM endpoints with live disputes**, I'm expecting this PR to result in a **4-5x improvement** in the dispute notice loading time:
	* the `/wc/v3/payments/charges/{charge_id}` fetch request takes **~6.5s** to respond (currently used for the disputed order notice).
	* the `/wc/v3/payments/disputes/{dispute_id}` fetch request takes **~1.5s** to respond (proposed to use in this PR).
* ~I've also added a `wcadmin_wcpay_single_order_view` tracks event to aid in tracking the "view disputed order" → "clicks disputed order notice button" funnel.~ removed – handled in #7437.
* This PR fixes #7076 since the `charge` endpoint is no longer utilised on the order edit screen.

**TODO:**
- [x] re-introduce `$dispute_id` argument to dispute webhook handler functions to add dispute id to order meta.
- [x] Remove tracks events from this PR (redundant, handled in #7437)



### Performance metrics

Method: Use [the browser `PerformanceElementTiming` API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceElementTiming) to measure the timestamp at which the disputed order notice is rendered. (see commit https://github.com/Automattic/woocommerce-payments/pull/7174/commits/d7ed392a1404d0f04c40eed80e0436cdae76d16a). Record results from 3x full-page refreshes with _Browser devtools -> Network -> Disable cache_ enabled.

For `develop` tests, I created a PR/branch with element performance timing measurements: #7175.

**On a live WooExpress site (Helix Demo site [Order 149](https://woo-delectably-sparkly-ninja.wpcomstaging.com/wp-admin/post.php?post=149&action=edit))**:

`develop`: average 6518ms
This branch average: 5632ms 
**13.5% improvement**

**On my local env with local WCPay-server running and proxied**:

`develop`: average 5375ms
This branch average: 4030ms 
**25% improvement**

### Testing instructions





* Create a disputed transaction using the card `4000000000000259` at checkout
* Navigate to the order edit screen
	* Ensure the disputed order notice is rendered
	* Ensure the tracks event `wcadmin_wcpay_single_order_view` is called with the properties `is_disputed: 1` & `dispute_status: needs_response`
		* Tip: check devtools network requests for `t.gif` with param `_en: wcadmin_wcpay_single_order_view`
	* Click the dispute notice and follow the steps to **Accept** the dispute
	* Ensure the tracks event `wcadmin_wcpay_single_order_view` is called with the properties `is_disputed: 1` & `dispute_status: lost`
		* Tip: check devtools network requests for `t.gif` with param `_en: wcadmin_wcpay_single_order_view`
* Navigate to the order edit screen for a non-disputed order
	* Ensure the disputed order notice is not rendered.
	* Ensure the tracks event `wcadmin_wcpay_single_order_view` is called with the properties `is_disputed: 0` & no `dispute_status`


-------------------


**Post merge**
